src/map/utils/synthutils.cpp

```diff
@@ -572,23 +501,59 @@ int32 doSynthSkillUp(CCharEntity* PChar)
                     }
                 }
 
+                // Cap skill gain if character hits the current cap
                 if((skillAmount + charSkill) > maxSkill)
                 {
                     skillAmount = maxSkill - charSkill;
                 }
 
+                uint16 skillCumulation = skillAmount;
+                uint8 skillHighest = 0;
+
+                if ((charSkill + skillAmount) > map_config.craft_common_cap) // If character is using the specialization system
+                {
+                    // Cycle through all skills except the one being used. Set the ID of the highest craft and the ammount of specialization points in use.
+                    for (int i = SKILL_WOODWORKING; i <= SKILL_COOKING; i++)
+                    {
+                        if (PChar->RealSkills.skill[i] > map_config.craft_common_cap)
+                        {
+                            skillCumulation += (PChar->RealSkills.skill[i] - map_config.craft_common_cap);
+                            if (skillID != i && PChar->RealSkills.skill[i] > skillHighest)
+                            {
+                                skillHighest = i;
+                            }
+                        }
+                    }
+                }
+
                 PChar->RealSkills.skill[skillID] += skillAmount;
+                if (skillHighest != 0 && skillCumulation > map_config.craft_specialization_points)
+                {
+                    PChar->RealSkills.skill[skillHighest] -= skillAmount;
+                }
+
                 PChar->pushPacket(new CMessageBasicPacket(PChar, PChar, skillID, skillAmount, 38));
 
-                if((charSkill/10) < (charSkill + skillAmount)/10)
+                if ((charSkill / 10) < (charSkill + skillAmount) / 10)
                 {
                     PChar->WorkingSkills.skill[skillID] += 0x20;
 
                     PChar->pushPacket(new CCharSkillsPacket(PChar));
-                    PChar->pushPacket(new CMessageBasicPacket(PChar, PChar, skillID, (charSkill + skillAmount)/10, 53));
+                    PChar->pushPacket(new CMessageBasicPacket(PChar, PChar, skillID, (charSkill + skillAmount) / 10, 53));
+                }
+
+                if (skillHighest != 0 && skillCumulation > map_config.craft_specialization_points && (PChar->RealSkills.skill[skillHighest] + skillAmount) / 10 > (PChar->RealSkills.skill[skillHighest]) / 10)
+                {
+                    PChar->WorkingSkills.skill[skillHighest] -= 0x20;
+                    PChar->pushPacket(new CCharSkillsPacket(PChar));
                 }
 
                 charutils::SaveCharSkills(PChar, skillID);
+
+                if (skillHighest != 0 && skillCumulation > map_config.craft_specialization_points)
+                {
+                    charutils::SaveCharSkills(PChar, skillHighest);
+                }
```

Comment:
Is restructuring possible so that all your `skillHighest != 0 && skillCumulation > map_config.craft_specialization_points` logic is contained in a single branch?

Reply:
Sure thing